xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/js/xwiki/actionbuttons/actionButtons.js

```diff
@@ -23,359 +23,518 @@
-      onSuccess : function(response) {
-        var progressOffset = response.responseJSON.progress.offset;
-        this.updateStatus(progressOffset);
-        if (progressOffset < 1) {
-          // Start polling for status updates, every second.
-          setTimeout(this.getStatus.bind(this, statusUrl, isContinue), 1000);
-        } else {
-          // Job complete.
-          this.progressBox.replace(this.savedBox);
-          this.maybeRedirect(isContinue);
+      if (isCreateFromTemplate) {
+        if (!isContinue) {
+          // Disable the form while waiting for the save&view operation to finish.
+          this.form.disable();
         }
-      }.bindAsEventListener(this),
-      on1223 : this.on1223.bindAsEventListener(this),
-      on0 : this.on0.bindAsEventListener(this),
-      onFailure : this.onFailure.bindAsEventListener(this)
-    });
-  }, 
-  updateStatus : function(progressOffset) {
-    var stringProgress = "" + (progressOffset * 100);
-    var dotIndex = stringProgress.indexOf('.');
-    if (dotIndex > -1) {
-      var stringProgress = stringProgress.substring(0, dotIndex + 2);
-    }
-    var newProgressBox = new XWiki.widgets.Notification(this.progressMessageTemplate.replace('__PROGRESS__', stringProgress), "inprogress");
-    this.progressBox.replace(newProgressBox);
-    this.progressBox = newProgressBox;
-  },
-  maybeRedirect : function(isContinue) {
-    var url = "";
-    if (!isContinue) {
-      // Redirect to view mode or to whatever URL was requested.
-      url = XWiki.currentDocument.getURL();
-      var xredirectElement = this.form.select('input[name="xredirect"]')[0];
-      if (xredirectElement && xredirectElement.value) {
-        url = xredirectElement.value;
+        // Start the progress display.
+        this.getStatus(response.responseJSON.links[0].href, isContinue);
+      } else if (isContinue) {
+        this.savingBox.replace(this.savedBox);
       }
-    } else if ($('body').hasClassName('previewbody')) {
-      // In preview mode, the &continue part of the save&continue should lead back to the edit action.
 
-      // Redirect to edit mode or to the previous edit mode, if not overridden.
-      url = XWiki.currentDocument.getURL('edit');
-      if (this.form.xcontinue && this.form.xcontinue.value) {
-        url = this.form.xcontinue.value;
+      // update the version
+      require(['xwiki-meta'], function (xm) {
+        xm.setVersion(response.responseJSON.newVersion);
+      });
+
+      // Announce that the document has been saved
+      // TODO: We should send the new version as a memo field
+      document.fire("xwiki:document:saved");
+    },
+    // 409 happens when the document is in conflict: i.e. someone's else edited it at the same time
+    on409 : function(response) {
+      this.progressBox.hide();
+      this.savingBox.hide();
+      this.savedBox.hide();
+      this.failedBox.hide();
+
+      var displayConflictModal = function (jsonAnswer) {
```

Comment:
You could also define this function as a method of the class, to be consistent. But I would not complain if you let it like this (it's more javascript-ish).